benchmarks/data/dataloader_benchmark.py

```diff
@@ -0,0 +1,322 @@
+
+This script measures:
+1. Dataloader initialization time
+2. Dataloading speed (time per batch)
+3. CPU memory utilization
+
+Usage:
+    python dataloader_benchmark.py --data_path /path/to/dataset --worker_method thread --batch_size 32 --num_workers 4
+"""
+import argparse
+import copy
+import gc
+import os
+import sys
+import threading
+import time
+
+import psutil
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import ConcatDataset
+from torchvision.models import resnet18
+
+
+def get_memory_usage(include_children=True):
+    """
+    Get current memory usage in MB.
+
+    Args:
+        include_children: If True, include memory of child processes
+
+    Returns:
+        Total memory usage in MB
+    """
+    process = psutil.Process()
+
+    main_memory = process.memory_full_info().pss
+
+    # If requested, add memory of all child processes
+    if include_children:
+        for child in process.children(recursive=True):
+            try:
+                child_mem = child.memory_full_info().pss
+                main_memory += child_mem
+            except (psutil.NoSuchProcess, psutil.AccessDenied, AttributeError):
+                # Process might have terminated or doesn't support PSS
+                try:
+                    # Fall back to USS
+                    print(f"Failed to get PSS for {child}, falling back to USS")
+                    child_mem = child.memory_info().uss
+                    main_memory += child_mem
+                except:
+                    pass
+
+    return main_memory / (1024 * 1024)
+
+
+def print_detailed_memory():
+    """Print detailed memory information."""
+    process = psutil.Process()
+    print("\nDetailed memory information:")
+    try:
+        print(
+            f"  USS (Unique Set Size): {process.memory_full_info().uss / (1024 * 1024):.2f} MB"
+        )
+        print(
+            f"  PSS (Proportional Set Size): {process.memory_full_info().pss / (1024 * 1024):.2f} MB"
+        )
+        print(
+            f"  RSS (Resident Set Size): {process.memory_info().rss / (1024 * 1024):.2f} MB"
+        )
+    except:
+        print("  Detailed memory info not available")
+
+
+def create_model():
+    """Create a simple model for benchmarking."""
+    model = resnet18()
+    return model
+
+
+def benchmark_dataloader(
+    dataset,
+    batch_size,
+    num_workers,
+    worker_method,
+    num_epochs=1,
+    max_batches=10,
+    multiprocessing_context=None,
+):
+    """Benchmark a dataloader with specific configuration."""
+    print(f"\n--- Benchmarking DataLoader with worker_method={worker_method} ---")
+
+    # Clear memory before starting
+    gc.collect()
+    torch.cuda.empty_cache()
+
+    # Measure memory before dataloader creation
+    memory_before = get_memory_usage()
```

Comment:
`memory_before` here is also before creating the model, so the memory increase is going to also include the size of the model.  Resnet is not huge so probably doesn't matter all that much.